27180 css nesting (#28473)
mdn/content: files/en-us/web/css/nesting_selector/index.md

## Syntax

```css
parent {
  /* parent style properties */
  & child {
    /* child style properties */
  }
}
```

### `&` nesting selector and whitespace

In this example nesting is done without the **`&` nesting selector**. The browser automatically adds whitespace.

```css
.parent-class {
  /* parent style properties */
  .child-class {
    /* child style properties */
  }
}
/* this becomes */
.parent-class {
  /* parent style properties */
}
.parent-class .child-class {
  /* child style properties */
}
```

When the nested selector needs to be attached to the parent selector, such as when using a {{cssxref('Pseudo-classes', 'pseudo selector')}} or creating [compound selectors](/en-US/docs/Web/CSS/CSS_selectors/Selector_structure#compound_selector) then this can be added immediately after the **`&` nesting selector**.

Without the **`&` nesting selector**, whitespace is added and, any child of the parent selector that is hovered will be styled.

```css
.parent-class {
  /* parent style properties */
  :hover {
    /* child style properties */
  }
}
/* this becomes */
.parent-class {
  /* parent style properties */
}
.parent-class :hover {
  /* child style properties */
}
```

> With the **`&` nesting selector** the parent will be styled when it is hovered.

Reviewer: ```suggestion
By including the **`&` nesting selector** we can get the desired rules after parsing, and the parent will be styled when hovered.
```

```css
.parent-class {
  /* parent style properties */
  &:hover {
    /* child style properties */
  }
}
/* this becomes */
.parent-class {
  /* parent style properties */
}
.parent-class:hover {
  /* child style properties */
}
```

## Appending the `&` nesting selector

The `&` nesting selector can also be appended to reverse the context of the selectors.

```css
.card {
  /* .card styles */
  .featured & {
    /* .featured .card styles */
  }
}
/* this becomes */
.card {
  /* .card styles */
}
.featured .card {
  /* .featured .card styles */
}
```

The `&` nesting selector can be placed multiple times:

```css
.card {
  /* .card styles */
  .featured & & & {
    /* .featured .card .card .card styles */
  }
}
/* this becomes */
.card {
  /* .card styles */
}
.featured .card .card .card {
  /* .featured .card .card .card styles */
}
```

## Examples

> Both of the following examples produce the same output, the first uses normal CSS styles and the second uses the `&` nesting selector.

Reviewer: ```suggestion
Both of the following examples produce the same output. The first one uses normal CSS styles and the second one uses the `&` nesting selector.
```

> ### Original CSS styles

Reviewer: ```suggestion
### Normal CSS styles
```

> This example uses original CSS styling.

Reviewer: ```suggestion
This example uses normal CSS styling.
```

#### HTML

```html
<p class="example">
  This paragraph <a href="#">contains a link</a>, try hovering or focusing it.
</p>
```

#### CSS

```css
.example {
  font-family: system-ui;
  font-size: 1.2rem;
}
.example > a {
  color: tomato;
}
.example > a:hover,
.example > a:focus {
  color: ivory;
  background-color: tomato;
}
```

#### Result

{{EmbedLiveSample('Original_CSS_styles','100%','50')}}

### Nested CSS styles

This example uses nested CSS styling.

#### HTML

```html
<p class="example">
  This paragraph <a href="#">contains a link</a>, try hovering or focusing it.
</p>
```

#### CSS

```css
.example {
  font-family: system-ui;
  font-size: 1.2rem;
  & a {
    color: tomato;
    &:hover,
    &:focus {
      color: ivory;
      background-color: tomato;
    }
  }
}
```

#### Result

> {{EmbedLiveSample('Nested_CSS_styles','100%','50')}}

Reviewer: Increase the height of the example iframes slightly, to get rid of the vertical scrollbars when rendered

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

> - [Using CSS nesting](/en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting)

Reviewer: Maybe link the main CSS Nesting page instead, so the reader can find all the related guides?

Author: Done

---
title: "& nesting selector"
slug: Web/CSS/Nesting_selector
page-type: css-selector
browser-compat: css.selectors.nesting
---

{{CSSRef}}

> The CSS **`&` nesting selector** is used to explicitly state the relationship between parent and child selectors when using [CSS Nesting](/en-US/docs/Web/CSS/CSS_nesting).

Reviewer 2: It's no parent and child __selectors__, it's parent and child **rules**. this is a huge different. We need to update this page to make that clear.

}
```

## Examples

Both of the following examples produce the same output. The first one uses normal CSS styles and the second one uses the `&` nesting selector.

### Normal CSS styles

This example uses normal CSS styling.

#### HTML

```html
<p class="example">
  This paragraph <a href="#">contains a link</a>, try hovering or focusing it.
</p>
```

#### CSS

```css
.example {
  font-family: system-ui;
  font-size: 1.2rem;
}

.example > a {
  color: tomato;
}

.example > a:hover,
.example > a:focus {
  color: ivory;
  background-color: tomato;
}
```

#### Result

{{EmbedLiveSample('Original_CSS_styles','100%','65')}}

### Nested CSS styles

This example uses nested CSS styling.

#### HTML

```html
<p class="example">
  This paragraph <a href="#">contains a link</a>, try hovering or focusing it.
</p>
```

#### CSS

```css
.example {
  font-family: system-ui;
  font-size: 1.2rem;
  & a {
    color: tomato;
    &:hover,
    &:focus {
      color: ivory;
      background-color: tomato;
    }
  }
}
```

#### Result

{{EmbedLiveSample('Nested_CSS_styles','100%','65')}}

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

> - [CSS nesting](/en-US/docs/Web/CSS/CSS_nesting) module

Reviewer 2: ```suggestion
- [Using CSS nesting](/en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting)
- [CSS nesting](/en-US/docs/Web/CSS/CSS_nesting) module
- [CSS selectors](/en-US/docs/Web/CSS/CSS_selectors) module
```

page-type: css-selector
browser-compat: css.selectors.nesting
---

{{CSSRef}}

The CSS **`&` nesting selector** explicitly states the relationship between parent and child rules when using [CSS Nesting](/en-US/docs/Web/CSS/CSS_nesting). Nested child rule selectors are relative to the parent rule selectors, with the child rule component having the same [specificity](/en-US/docs/Web/CSS/CSS_nesting/Nesting_and_specificity) weight as if they were within {{cssxref(":is", ":is()")}}.

## Syntax

```css
parentRule {
  /* parent rule style properties */
  & childRule {
    /* child rule style properties */
  }
}
```

### `&` nesting selector and whitespace

In this example, nesting is done without the **`&` nesting selector**. When the browser parses the nested selectors, it automatically adds whitespace between the selectors to create a new CSS selector rule.

````css
.parent-rule {
  /* parent rule properties */
  .child-rule {
    /* child rule properties */
  }
}

/* the browser parses this as */
.parent-rule {
  /* parent rule style properties */
}

.parent-rule .child-rule {
  /* style properties for .child-rule descendants for .parent-rule ancestors */

> }

Reviewer 2: ```suggestion
}
```
```

Reviewer 2: the code block is missing the closing ticks

Author: Done

> When the nested rule needs to be attached (with no whitespace) to the parent rule, such as when using a {{cssxref('Pseudo-classes', 'pseudo class')}} or creating [compound selectors](/en-US/docs/Web/CSS/CSS_selectors/Selector_structure#compound_selector), the **`&` nesting selector** can be immediately prepended to achieve the desired effect.

Author: Yes I shall update this.

> When the nested rule needs to be attached (with no whitespace) to the parent rule, such as when using a {{cssxref('Pseudo-classes', 'pseudo class')}} or creating [compound selectors](/en-US/docs/Web/CSS/CSS_selectors/Selector_structure#compound_selector), the **`&` nesting selector** can be immediately prepended to achieve the desired effect.

Author: Updated this
